src/ecdsa/keys.py

```diff
@@ -870,30 +873,36 @@ def from_pem(cls, string, hashfunc=sha1):
         :return: Initialised SigningKey object
         :rtype: SigningKey
         """
-        # the privkey pem may have multiple sections, commonly it also has
-        # "EC PARAMETERS", we need just "EC PRIVATE KEY".
         if not PY2 and isinstance(string, str):
             string = string.encode()
-        privkey_pem = string[
-            string.index(b("-----BEGIN EC PRIVATE KEY-----")) :
-        ]
-        return cls.from_der(der.unpem(privkey_pem), hashfunc)
+
+        # the privkey pem may have multiple sections, commonly it also has
+        # "EC PARAMETERS", we need just "EC PRIVATE KEY".
+        ec_private_key_index = string.find(b"-----BEGIN EC PRIVATE KEY-----")
+        if ec_private_key_index != -1:
+            return cls.from_der(
+                der.unpem(string[ec_private_key_index:]), hashfunc, pkcs8=False
+            )
+
+        private_key_index = string.find(b"-----BEGIN PRIVATE KEY-----")
+        if private_key_index != -1:
+            return cls.from_der(
+                der.unpem(string[private_key_index:]), hashfunc, pkcs8=True
+            )
```

Comment:
that feels to me like a lot of code duplication just to set a single variable

Reply:
This will be cleared up once `from_der` detects the incoming format

Reply:
yes and no, I think we still should abort on `BEGIN RSA PRIVATE KEY` and such
also if we have option to force the format, not only the inside DER but the outside PEM needs to match the expected format